portability-core/src/main/java/org/dataportabilityproject/cloud/google/GooglePersistentKeyValueStore.java

```diff
@@ -52,8 +57,69 @@ public GooglePersistentKeyValueStore(Datastore datastore) {
 
   @Override
   public void put(String key, Map<String, Object> data) throws IOException {
-    Entity.Builder builder = Entity.newBuilder(getKey(key))
-        .set(CREATED_FIELD, Timestamp.now());
+    Entity entity = createEntity(key, data);
+    datastore.put(entity);
+  }
+
+  @Override
+  public Map<String, Object> get(String key) {
+    Entity entity = datastore.get(getKey(key));
+    return getProperties(entity);
+  }
+
+  @Override
+  public String getFirst(String prefix) {
+    Query<Key> query = Query.newKeyQueryBuilder().setKind(KIND)
+        .build();
+    QueryResults<Key> results = datastore.run(query);
+    while(results.hasNext()) {
+      String key = results.next().getName();
+      if(key.startsWith(prefix)) {
+        return key;
+      }
+    }
+    return null;
+  }
+
+  @Override
+  public void delete(String key) {
+    datastore.delete(getKey(key));
+  }
+
+  @Override
+  public boolean atomicUpdate(String previousKeyStr, String newKeyStr, Map<String, Object> data) {
+    Transaction transaction = datastore.newTransaction();
+
+    try {
+      Key previousKey = getKey(previousKeyStr);
+      Entity previousEntity = transaction.get(previousKey);
+      if (previousEntity == null) {
+        logger.debug("Could not find previous key {}", previousKeyStr);
+        return false;
+      }
+
+      Key newKey = getKey(newKeyStr);
+      Entity newEntity = transaction.get(newKey);
+      if (newEntity != null) {
+        logger.debug("Updated key already exists: {}", newKeyStr);
```

Comment:
Do we have to do anything with the transaction here?

Reply:
same here